k8s/k8s.go

```diff
@@ -1137,50 +1160,162 @@ func (k *k8sOps) isPVCShared(pvc *v1.PersistentVolumeClaim) bool {
 
 // PVCs APIs - END
 
+// Snapshot APIs - BEGIN
+
+func (k *k8sOps) CreateSnapshot(snap *snap_v1.VolumeSnapshot) (*snap_v1.VolumeSnapshot, error) {
+	if err := k.initK8sClient(); err != nil {
+		return nil, err
+	}
+	var result snap_v1.VolumeSnapshot
+	if err := k.snapClient.Post().
+		Name(snap.Metadata.Name).
+		Resource(snap_v1.VolumeSnapshotResourcePlural).
+		Namespace(snap.Metadata.Namespace).
+		Body(snap).
+		Do().Into(&result); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func (k *k8sOps) DeleteSnapshot(snap *snap_v1.VolumeSnapshot) error {
+	if err := k.initK8sClient(); err != nil {
+		return err
+	}
+	return k.snapClient.Delete().
+		Name(snap.Metadata.Name).
+		Resource(snap_v1.VolumeSnapshotResourcePlural).
+		Namespace(snap.Metadata.Namespace).
+		Do().Error()
+}
+
+func (k *k8sOps) ValidateSnapshot(snap *snap_v1.VolumeSnapshot) error {
+	if err := k.initK8sClient(); err != nil {
+		return err
+	}
+	t := func() (interface{}, error) {
+		status, err := k.GetSnapshotStatus(snap)
+		if err != nil {
+			return "", true, err
+		}
+
+		for _, condition := range status.Conditions {
+			if condition.Type == snap_v1.VolumeSnapshotConditionReady && condition.Status == v1.ConditionTrue {
+				return "", true, nil
+			} else if condition.Type == snap_v1.VolumeSnapshotConditionError && condition.Status == v1.ConditionTrue {
+				return "", false, &ErrSnapshotFailed{
+					ID:    snap.Metadata.Name,
+					Cause: fmt.Sprintf("Snapshot Status %v", status),
+				}
+			}
+		}
+
+		return "", true, &ErrSnapshotNotReady{
+			ID:    snap.Metadata.Name,
+			Cause: fmt.Sprintf("Snapshot Status %v", status),
+		}
+	}
+
+	if _, err := task.DoRetryWithTimeout(t, 5*time.Minute, 10*time.Second); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k *k8sOps) GetVolumeForSnapshot(snap *snap_v1.VolumeSnapshot) (string, error) {
+	if err := k.initK8sClient(); err != nil {
+		return "", err
+	}
+
+	var result snap_v1.VolumeSnapshot
+	if err := k.snapClient.Get().
+		Name(snap.Metadata.Name).
+		Resource(snap_v1.VolumeSnapshotResourcePlural).
+		Namespace(snap.Metadata.Namespace).
+		Do().Into(&result); err != nil {
+		return "", err
+	}
+	return result.Metadata.Name, nil
```

Comment:
Is the snapshot name same as volume name? If yes, why do we call the method getVolumeForSnapshot instead of getSnapshotName or directly use the snapshot.metadata.name wherever we are using it?

Reply:
It doesn't have to be the same. Right now stork uses the snapshot name, which could change. This creates an abstraction layer.